File: examples/convert_json_to_nyml.js
#!/usr/bin/env node

/**
 * Convert JSON to NYML format
 */

function jsonToNyml(obj, indent = 0) {
  const lines = [];
  const indentStr = '  '.repeat(indent);

  if (typeof obj === 'object' && obj !== null) {
    if (Array.isArray(obj)) {
      // Array - convert to multiline string
      lines.push(`${indentStr}|`);
      for (const item of obj) {
        lines.push(`${indentStr}  ${item}`);
      }
      if (obj.length > 0) {
        lines.push(''); // Empty line after multiline
      }
    } else {
      // Object
      for (const [key, value] of Object.entries(obj)) {
        // Quote key if it contains special characters or spaces
        const keyStr = needsQuoting(key) ? `"${key}"` : key;

        if (typeof value === 'object' && value !== null) {
          if (Array.isArray(value)) {
            // Array - convert to multiline string
            lines.push(`${indentStr}${keyStr}: |`);
            for (const item of value) {
              lines.push(`${indentStr}  ${item}`);
            }
            if (value.length > 0) {
              lines.push(''); // Empty line after multiline
            }
          } else {
            // Nested object
            lines.push(`${indentStr}${keyStr}:`);
            lines.push(...jsonToNyml(value, indent + 1));
          }
        } else if (typeof value === 'string' && value.includes('\n')) {
          // Multiline string - convert back to | syntax
          lines.push(`${indentStr}${keyStr}: |`);
          const contentLines = value.split('\n');
          // Remove trailing empty line if present
          if (contentLines.length > 0 && contentLines[contentLines.length - 1] === '') {
            contentLines.pop();
          }
          for (const line of contentLines) {
            lines.push(`${indentStr}  ${line}`);
          }
          lines.push(''); // Empty line after multiline
        } else {
          // Simple value
          let valueStr = String(value);
          // Quote value if it contains special characters
          if (needsQuotingValue(valueStr)) {
            valueStr = `"${valueStr}"`;
          }
          lines.push(`${indentStr}${keyStr}: ${valueStr}`);
        }
      }
    }
  } else {
    // Primitive value
    lines.push(`${indentStr}value: ${obj}`);
  }

  return lines;
}

function needsQuoting(key) {
  // Quote if contains spaces, colons, special chars, or starts with digit
  return (key.includes(' ') ||
          key.includes(':') ||
          key.includes('@') ||
          key.startsWith('"') ||
          key.startsWith('#') ||
          /^\d/.test(key) ||
          !/^[a-zA-Z0-9._-]+$/.test(key));
}

function needsQuotingValue(value) {
  // Quote if contains spaces, quotes, or special characters
  return (value.includes(' ') ||
          value.includes('"') ||
          value.includes("'") ||
          /[#@:$]/.test(value) ||
          (/^\d/.test(value) && /[-.]/.test(value)));
}

// Main execution
if (require.main === module) {
  const fs = require('fs');

  if (process.argv.length !== 3) {
    console.error('Usage: node convert_json_to_nyml.js <json_file>');
    process.exit(1);
  }

  const jsonFile = process.argv[2];

  try {
    const data = JSON.parse(fs.readFileSync(jsonFile, 'utf8'));
    const lines = jsonToNyml(data);
    console.log(lines.join('\n'));
  } catch (error) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
}

module.exports = { jsonToNyml, needsQuoting, needsQuotingValue };

File: examples/convert_json_to_nyml.py
#!/usr/bin/env python3
"""Convert JSON to NYML format."""

import json
import sys

def json_to_nyml(obj, indent=0):
    """Convert a JSON object to NYML format."""
    lines = []
    indent_str = '  ' * indent

    if isinstance(obj, dict):
        for key, value in obj.items():
            # Quote key if it contains special characters or spaces
            if needs_quoting(key):
                key_str = f'"{key}"'
            else:
                key_str = key

            if isinstance(value, dict):
                # Nested object
                lines.append(f'{indent_str}{key_str}:')
                lines.extend(json_to_nyml(value, indent + 1))
            elif isinstance(value, list):
                # Array - convert to multiline string
                lines.append(f'{indent_str}{key_str}: |')
                for item in value:
                    lines.append(f'{indent_str}  {item}')
                # Add empty line after multiline
                if value:  # Only if array is not empty
                    lines.append('')
            elif isinstance(value, str) and '\n' in value:
                # Multiline string - convert back to | syntax
                lines.append(f'{indent_str}{key_str}: |')
                # Split by lines and add proper indentation
                content_lines = value.split('\n')
                # Remove trailing empty line if present (multiline strings end with \n)
                if content_lines and content_lines[-1] == '':
                    content_lines.pop()
                
                for line in content_lines:
                    lines.append(f'{indent_str}  {line}')
                lines.append('')  # Empty line after multiline
            else:
                # Simple value
                value_str = str(value)
                # Quote value if it contains special characters
                if needs_quoting_value(value_str):
                    value_str = f'"{value_str}"'
                lines.append(f'{indent_str}{key_str}: {value_str}')

    elif isinstance(obj, list):
        # Top-level array (shouldn't normally happen, but handle it)
        for i, item in enumerate(obj):
            key = f'item{i}'
            if isinstance(item, (dict, list)):
                lines.append(f'{indent_str}{key}:')
                lines.extend(json_to_nyml(item, indent + 1))
            else:
                lines.append(f'{indent_str}{key}: {item}')

    else:
        # Top-level primitive (shouldn't normally happen)
        lines.append(f'{indent_str}value: {obj}')

    return lines

def needs_quoting(key):
    """Check if a key needs to be quoted."""
    # Quote if contains spaces, colons, special chars, or starts with digit
    return (' ' in key or
            ':' in key or
            '@' in key or
            key.startswith(('"', '#')) or
            key[0].isdigit() or
            not key.replace('_', '').replace('-', '').replace('.', '').isalnum())

def needs_quoting_value(value):
    """Check if a value needs to be quoted."""
    # Always quote strings that contain spaces, quotes, or special characters
    # Also quote if it looks like a number but contains special chars
    return (' ' in value or 
            '"' in value or 
            "'" in value or
            any(char in value for char in '#@:$') or
            (value.replace('.', '').replace('-', '').isdigit() and ('.' in value or '-' in value)))

def main():
    if len(sys.argv) != 2:
        print("Usage: python convert_json_to_nyml.py <json_file>")
        sys.exit(1)

    json_file = sys.argv[1]

    try:
        with open(json_file, 'r') as f:
            data = json.load(f)

        lines = json_to_nyml(data)
        print('\n'.join(lines))

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()